crates/task-impls/src/quorum_proposal_recv/handlers.rs

```diff
@@ -0,0 +1,232 @@
+    proposal: &Proposal<TYPES, QuorumProposal<TYPES>>,
+    event_sender: &Sender<Arc<HotShotEvent<TYPES>>>,
+    justify_qc: &QuorumCertificate<TYPES>,
+    task_state: &mut QuorumProposalRecvTaskState<TYPES, I>,
+) -> Option<QuorumProposal<TYPES>> {
+    let view_number = proposal.data.view_number();
+    let mut consensus_write = task_state.consensus.write().await;
+
+    warn!(
+        "Proposal's parent missing from storage with commitment: {:?}",
+        justify_qc.data.leaf_commit
+    );
+    let leaf = Leaf::from_quorum_proposal(&proposal.data);
+
+    let state = Arc::new(
+        <TYPES::ValidatedState as ValidatedState<TYPES>>::from_header(&proposal.data.block_header),
+    );
+    let view = View {
+        view_inner: ViewInner::Leaf {
+            leaf: leaf.commit(),
+            state,
+            delta: None,
+        },
+    };
+
+    consensus_write.update_validated_state_map(view_number, view.clone());
+    consensus_write.update_saved_leaves(leaf.clone());
+
+    if let Err(e) = task_state
+        .storage
+        .write()
+        .await
+        .update_undecided_state(
+            consensus_write.saved_leaves().clone(),
+            consensus_write.validated_state_map().clone(),
+        )
+        .await
+    {
+        warn!("Couldn't store undecided state.  Error: {:?}", e);
+    }
+
+    let liveness_check = justify_qc.view_number() > consensus_write.locked_view();
+
+    let high_qc = consensus_write.high_qc().clone();
+    drop(consensus_write);
+
+    // Broadcast that we've updated our consensus state so that other tasks know it's safe to grab.
+    broadcast_event(
+        HotShotEvent::ValidatedStateUpdated(view_number, view).into(),
+        event_sender,
+    )
+    .await;
+    broadcast_event(
+        HotShotEvent::NewUndecidedView(leaf.clone()).into(),
+        event_sender,
+    )
+    .await;
+
+    let mut current_proposal = None;
+    if liveness_check {
+        current_proposal = Some(proposal.data.clone());
+
+        let new_view = proposal.data.view_number + 1;
+
+        // This is for the case where we form a QC but have not yet seen the previous proposal ourselves
+        let should_propose = task_state.quorum_membership.leader(new_view) == task_state.public_key
+            && high_qc.view_number == current_proposal.clone().unwrap().view_number;
+
+        if should_propose {
+            debug!(
+                "Attempting to publish proposal after voting for liveness; now in view: {}",
+                *new_view
+            );
+            broadcast_event(
+                HotShotEvent::LivenessCheckProposalRecv(proposal.data.clone()).into(),
+                event_sender,
+            )
+            .await;
+        }
+    }
+
+    // If we are missing the parent from storage, the safety check will fail.  But we can
+    // still vote if the liveness check succeeds.
+    current_proposal
+}
+
+/// Handles the `QuorumProposalRecv` event.
```

Comment:
Done, however I elected to keep the parameter until we find a better solution, here's why: right now we take the proposal as input, and follow one of two `Ok` paths depending on if we need to submit the liveness proposal. Because of this, we'd still need to return something to signify that we're performing the liveness check and, as a result, I'd have to make some new descriptive variable, or use the data itself which I think is more descriptive. 

In regard to your above comment, I think that simplifying this to remove the branching cases would be definitely useful, but I am hesitant to over-design this implementation as we'd end up spending considerable effort making it perfect, so I'd like to opt for iterative refinement when we've had the chance to the let dependency tasks settle in production.

This is all with a grain of salt. If you happen to have some ideas I'm more than happy and willing to get them in right away, my qualm is just that I don't yet know of an explicitly cleaner way to do this without first implementing the state update changes that I wanted to do (but that'd require updating nearly every consensus and consensus-adjacent task).

Reply:
I think it's fine to have both `Ok(Some)` and `Ok(None)`, but if we are going to keep this return type, in the consensus task, I think we should call `spawn_vote_task` for both cases. Currently, we only spawn the vote task for the `Ok(Some)` case.

Btw, looks like the documentation for the return type isn't added yet.

Reply:
> I think it's fine to have both `Ok(Some)` and `Ok(None)`, but if we are going to keep this return type, in the consensus task, I think we should call `spawn_vote_task` for both cases. Currently, we only spawn the vote task for the `Ok(Some)` case.

Oh, I just realized that we had `handle_quorum_proposal_recv` defined in both the consensus and the quorum proposal recv tasks, so the changes here wouldn't affect the consensus task. Can we make the one in the consensus task `pub(crate)` to avoid misuse?